### DISCUSSION TITLE
How to customize my own PoS reward mechanism 🤔

### DISCUSSION THREAD

**Discussion by eddieran** [0xPolygon/polygon-edge#667] [Q&A]
### Description

We want to use Edge to make our own chain, PoS, and can customize our own Tokenomics. during this process, I found some doubts and would like to ask questions here to get answers. 

### Q2. Edge would `preDeploy` staking contract every time it builds block (if there is exist staking contract at that address, how would re-deploying effect state of blockchain? 
![image](https://user-images.githubusercontent.com/8403607/181894041-d0086b0b-f71b-4e9d-b722-873b8ff146e3.png)
<img width="856" alt="image" src="https://user-images.githubusercontent.com/8403607/181894409-8d608b05-8042-4f38-b386-1720694612b8.png">
If we have already deployed the staking contract, why do we need to deploy it again during each PreStateCommit? What state changes will this cause?

### Q3. `AddSealingReward` is not called anywhere, how edge reward proposer? 

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/8403607/181895231-48ee681d-ad9b-4f47-b753-0b6e83b81b7b.png">
I looked at the implementation and logic in `ibft.go` and `pos.go` and didn't find any place to redirect block proposer, and found `AddSealingReward` in the whole repository, but it was only called in the test cases.
Is it possible that the current PoS implementation does not reward miners?
If we want to customize our own PoS reward mechanism, where do we suggest to start?

#### **Any replies and suggestions from the developers would be greatly appreciated.**

**Comment by laviniat1996** [comment]
Hi @SureEdding ,

Thank you for your questions and sorry for the delayed reply.

The only reward mechanism Polygon Edge has, which works for both PoA and PoS, is implemented in the executor.go file (from the State folder). The proposer is going to get the coinbase fee. The implementation is:
https://github.com/0xPolygon/polygon-edge/blob/d3d4f8ac4f6b401c01e0e8b6afef3d78aa58d8e5/state/executor.go#L486-L488

Regarding your second question, that function is part of a hook that is only called when switching from PoA to PoS:
https://github.com/0xPolygon/polygon-edge/blob/develop/consensus/ibft/pos.go#L48

**Reply by eddieran** [reply]
thank you for repling!
 so the proposer can get all the gasUsed as reward in each block? 🤔 
so if we want to customize the reward mechanism, we can try to add our own formula to decide the reward amount in apply function?
